hr_timesheet_sheet/models/hr_timesheet_sheet.py

```diff
@@ -0,0 +1,621 @@
+
+
+class Sheet(models.Model):
+    _name = 'hr_timesheet.sheet'
+    _inherit = ['mail.thread', 'mail.activity.mixin', 'portal.mixin']
+    _table = 'hr_timesheet_sheet'
+    _order = "id desc"
+    _description = 'Timesheet Sheet'
+
+    def _default_date_start(self):
+        user = self.env['res.users'].browse(self.env.uid)
+        r = user.company_id and user.company_id.sheet_range or WEEKLY
+        today = fields.Date.from_string(fields.Date.context_today(self))
+        if r == WEEKLY:
+            return today - relativedelta(days=today.weekday())
+        elif r == MONTHLY:
+            return fields.Date.date(today.year, today.month, 1)
+        elif r == YEARLY:
+            return fields.Date.date(today.year, 1, 1)
+        return today
+
+    def _default_date_end(self):
+        user = self.env['res.users'].browse(self.env.uid)
+        r = user.company_id and user.company_id.sheet_range or WEEKLY
+        today = fields.Date.from_string(fields.Date.context_today(self))
+        if r == WEEKLY:
+            return today + relativedelta(days=6-today.weekday())
+        elif r == MONTHLY:
+            return today + relativedelta(months=1, day=1, days=-1)
+        elif r == YEARLY:
+            return today + relativedelta(years=1, month=1, day=1, days=-1)
+        return today
+
+    def _default_employee(self):
+        emp_ids = self.env['hr.employee'].search(
+            [('user_id', '=', self.env.uid)])
+        return emp_ids and emp_ids[0] or False
+
+    name = fields.Char(
+        string="Note",
+        states={'confirm': [('readonly', True)], 'done': [('readonly', True)]},
+    )
+    employee_id = fields.Many2one(
+        comodel_name='hr.employee',
+        string='Employee',
+        default=_default_employee,
+        required=True,
+    )
+    user_id = fields.Many2one(
+        comodel_name='res.users',
+        related='employee_id.user_id',
+        string='User',
+        store=True,
+        readonly=True,
+    )
+    date_start = fields.Date(
+        string='Date From',
+        default=_default_date_start,
+        required=True,
+        index=True,
+        states={'new': [('readonly', False)]},
+    )
+    date_end = fields.Date(
+        string='Date To',
+        default=_default_date_end,
+        required=True,
+        index=True,
+        states={'new': [('readonly', False)]},
+    )
+    timesheet_ids = fields.One2many(
+        comodel_name='account.analytic.line',
+        inverse_name='sheet_id',
+        string='Timesheets',
+        readonly=True,
+        states={
+            'draft': [('readonly', False)],
+            'new': [('readonly', False)]},
+    )
+    line_ids = fields.One2many(
+        comodel_name='hr_timesheet.sheet.line',
+        compute='_compute_line_ids',
+        string='Timesheets',
+        states={
+            'draft': [('readonly', False)],
+            'new': [('readonly', False)]},
+    )
+    state = fields.Selection([
+        ('new', 'New'),
```

Comment:
Consider to remove the item 'new' of the state field.
Actually the timesheet sheet is created with state field as 'new' by default, but it automatically goes to 'draft' while created (see `def create()`). Then 'new' is never used again ... so I would say that state 'new' is completely useless.

Reply:
the difference is that in <code>new</code>, you select the dates. In <code>draft</code> you can't change the dates.

Reply:
Ok, then for me it's better to keep this part as it is. Thank you for checking!

Reply:
Well, I finally deleted this state.